File: GFG POTD/007-July-2024/002-linked-list-of-strings-forms-a-palindrome.java
// https://www.geeksforgeeks.org/problems/linked-list-of-strings-forms-a-palindrome/1



//{ Driver Code Starts
import java.util.*;

class Node {
    String data;
    Node next;

    Node(String x) {
        data = x;
        next = null;
    }
}

class LinkedList {
    Node head;

    LinkedList() { head = null; }

    void addNode(String str) {
        if (head == null) {
            head = new Node(str);
        } else {
            Node temp = head;
            while (temp.next != null) {
                temp = temp.next;
            }
            temp.next = new Node(str);
        }
    }

    void print() {
        Node temp = head;
        while (temp != null) {
            System.out.println(temp.data);
            temp = temp.next;
        }
    }
}

public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int T = sc.nextInt();

        while (T-- > 0) {
            int K = sc.nextInt();
            LinkedList list = new LinkedList();

            for (int i = 0; i < K; i++) {
                String str = sc.next();
                list.addNode(str);
            }

            Solution ob = new Solution();
            boolean ans = ob.compute(list.head);
            if (ans) {
                System.out.println("true");
            } else {
                System.out.println("false");
            }
        }
        sc.close();
    }
}
// } Driver Code Ends

// User function Template for Java
class Solution {
    public boolean compute(Node root) {
        StringBuilder store = new StringBuilder();
        getString(root, store);
        
        return isPalindrome(store.toString());
    }

    // TC : O(N)
    // SC : O(1)
    private static void getString(Node root, StringBuilder store) {
        while(root != null) {
            store.append(root.data);
            root = root.next;
        }
    }

    // TC : O(N)
    // SC : O(1)
    private static boolean isPalindrome(String store) {
        int i = 0, j = store.length() - 1;
        while (i <= j) {
            if (store.charAt(i) != store.charAt(j)) {
                return false;
            }
            i ++;
            j --;
        }
        return true;
    }
}

File: GFG POTD/007-July-2024/002-linked-list-of-strings-forms-a-palindrome.cpp
// https://www.geeksforgeeks.org/problems/linked-list-of-strings-forms-a-palindrome/1



//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;

struct Node {
    string data;
    Node *next;

    Node(string x) {
        data = x;
        next = NULL;
    }
};

Node *newNode(string str) {
    Node *new_node = new Node(str);
    new_node->next = NULL;
    return new_node;
}

void print(Node *root) {
    Node *temp = root;

    while (temp != NULL) {
        cout << temp->data << endl;
        temp = temp->next;
    }
}
// } Driver Code Ends
/*
The structure of linked list is the following
struct Node
{
    string data;
    Node* next;

    Node(int x){
        data = x;
        next = NULL;
    }
};
*/
class Solution {
  private:
    // TC : O(N)
    // SC : O(1)
    void getString(Node* head, string& store) {
        while(head) {
            store += (head -> data);
            head   = head -> next;
        }
    }

    // TC : O(N)
    // SC : O(1)
    bool isPalindrome(string& store) {
        int i = 0, j = store.length() - 1;
        while (i <= j) {
            if (store[i] != store[j]) {
                return false;
            }
            i ++, j --;
        }
        return true;
    }
  public:
    // TC : O(N)
    // SC : O(N)
    bool compute(Node* head) {
        string store;
        getString(head, store);
        
        return isPalindrome(store);
    }
};

//{ Driver Code Starts.
int main() {
    // your code goes here
    int T;
    cin >> T;

    while (T--) {
        int K;
        cin >> K;
        struct Node *head = NULL;
        struct Node *temp = head;

        for (int i = 0; i < K; i++) {
            string str = "";
            cin >> str;
            if (head == NULL)
                head = temp = new Node(str);
            else {
                temp->next = new Node(str);
                temp = temp->next;
            }
        }
        Solution ob;
        bool ans = ob.compute(head);
        if (ans)
            cout << "true" << endl;
        else
            cout << "false" << endl;
    }
    return 0;
}
// } Driver Code Ends